### ISSUE TITLE
Bear Improvement: Use `count lines of code(cloc)` in LineCountBear to enhance usability.

### ISSUE THREAD

**Issue by sahilarora535** [coala/coala-bears#1420]
Coala does not have `cloc` bear. `cloc` is a command line utility which counts lines of code in a directory/file. This is the output when run on this repository:
```
     352 text files.
     352 unique files.                                          
     422 files ignored.

http://cloc.sourceforge.net v 1.60  T=1.75 s (176.2 files/s, 11666.3 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                         237           2785           4633          10765
C                               32            169            105            959
YAML                             7             28             24            241
Bourne Shell                     7             48             42            190
make                             1             29              6            181
XML                              8              0              0             59
C/C++ Header                     2             15             20             43
Java                             4              7              2             28
CSS                              2              3              0             18
XSD                              1              2              0             16
C#                               2              0              0             14
DTD                              1              0              0             10
Go                               2              1              0              8
PHP                              2              0              0              6
Javascript                       1              0              0              3
-------------------------------------------------------------------------------
SUM:                           309           3087           4832          12541
-------------------------------------------------------------------------------
```

**Comment by Makman2** [comment]
Is https://github.com/coala/coala-bears/blob/master/bears/general/LineCountBear.py what you look for?

**Comment by sahilarora535** [comment]
@Makman2 Actually `cloc` is a bit different. It, as the name suggests `count lines of code`, counts the lines of code per file, and gives a detail of code lines, comments, blank lines per language. The `LineCountBear` just count lines of code, without statistics. See the [first comment](https://github.com/coala/coala-bears/issues/1420#issue-206232264) on this issue for reference.

**Comment by Makman2** [comment]
usecase?

**Comment by sahilarora535** [comment]
@Makman2 Counting lines of code! I can't think of a better usecase. We have `LineCountBear`, when it can be simply done using `cat * | wc -l`, we should have this as well. What's say?

**Comment by Makman2** [comment]
Thing is the `LineCountBear` does support a limit: https://github.com/coala/coala-bears/blob/master/bears/general/LineCountBear.py#L18

Not sure if it makes sense to restrict the complete amount of code from a specific language^^

**Comment by meetmangukiya** [comment]
I guess the limit forces the developer to make more modular and readable code

**Comment by Makman2** [comment]
Yeah but this does not really work when you say "please write only 1000 LOC for python in your complete project" :3 That's what I wanted to say :)

**Comment by underyx** [comment]
@Makman2 this would be useful: currently, we have a 1000 line limit on files, but with cloc, we could change it to like 500 actual lines of code, which would exclude blank lines, docstrings, comments, etc.

**Comment by Makman2** [comment]
Shouldn't we then improve our bear by using cloc?

**Comment by underyx** [comment]
That's also a possibility. It would also count hacks like `import os; os.setenv('FOO', 'BAR')` as two lines then.

**Comment by Makman2** [comment]
sounds good. Let's improve our current bear :+1:

**Comment by meetmangukiya** [comment]
> sounds good. Let's improve our current bear :+1:

Lets change the issue title, then?

**Comment by sahilarora535** [comment]
@underyx  @Makman2 @meetmangukiya title changed. This will improve the usability of the bear for sure. :smile:

**Comment by underyx** [comment]
I'd suggest leaving `max_lines_per_file` alone and adding `cloc` count checking only with a new `max_lines_of_code_per_file` option.

**Comment by Makman2** [comment]
maybe we just have a flag `count statements=True/False`  (or similar) which then kicks in `cloc` or keeps the simple behaviour?

**Comment by dracarys09** [comment]
unassigning myself because @sahilarora535 already made some progress with this and I'm a little busy with my exams and other issues...

**Comment by jayvdb** [comment]
The current LineCountBear works for any file type, and without any external dependency.
It is basic, but it works well for a *very* large number of scenarios.

Bear metadata does not have the ability to have optional dependencies, so until that is fixed, it would be a **breaking change** to add *any* external dependency to `LineCountBear`.  It would break existing CI, because the new dependency wont be installed by their CI scripts.  Hence changing this to being a new bear proposal instead of being about `LineCountBear`.

LOC algorithms are hard to implement properly, and are language specific to do properly.
We have some fairly generic functionality with the coalib languages metadata, but it is in its infancy, with very few languages supported, and very few statement types understood, and a extremely small test suite not covering even rudimentary constructs in Python even.  Any LOC algorithm based on that would have such gaping holes as to cause embarrassment, to be honest.

An achievable enhancement with `LineCountBear` is excluding blank lines https://github.com/coala/coala-bears/issues/1478 .  The next step would be excluding comments.  But even that is *complex*.  What is a comment?  Our coalang information is very basic in that regard. And then, many languages have comments which should count as code, as they contain vital elements and the code will fail if those comments are removed.

So, if we want a LOC bear, either 1) use an existing external tool, or 2) create lots of enhancements for the coalang section of coala, or maybe 3) create a language specific native bear .

**Comment by sahilarora535** [comment]
@jayvdb There actually is a tool for counting lines of code, which is `cloc`. I agree that implementing LOC is a very difficult task, however, there are already tools for that. The first comment in this thread(https://github.com/coala/coala-bears/issues/1420#issue-206232264) has been produced by the tool [`cloc`](https://github.com/AlDanial/cloc). And you are right, maybe we should not play with the current `LineCountBear`, rather make a new one.
